src/rsa/blinding.rs

```diff
@@ -12,52 +12,127 @@
 // OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
 // CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 
-use error;
-use super::bigint;
+//! RSA base blinding.
+//!
+//! This is based on the suggestions in the paper
+//! "Timing Attacks on Implementations of Diffie-Hellman, RSA, DSS, and Other
+//! Systems" by Paul C. Kocher.
 
-pub struct Blinding(*mut BN_BLINDING);
+use {error, rand};
+use core;
+use super::{bigint, N};
 
-impl Drop for Blinding {
-    fn drop(&mut self) { unsafe { GFp_BN_BLINDING_free(self.as_mut_ref()) } }
-}
+pub struct Blinding(Option<Contents>);
 
-// `Blinding` uniquely owns and references its contents.
-unsafe impl Send for Blinding {}
+struct Contents {
+    blinding_factor: bigint::Elem<N>, // `(1 / v_i)**e` from the paper.
+    blinding_factor_inv: bigint::Elem<N>, // `v_f` from the paper.
+    remaining: usize,
+}
 
 impl Blinding {
-    pub fn new() -> Result<Blinding, error::Unspecified> {
-        let r = unsafe { GFp_BN_BLINDING_new() };
-        if r.is_null() {
-            return Err(error::Unspecified);
-        }
-        Ok(Blinding(r))
+    pub fn new() -> Self { Blinding(None) }
+
+    pub fn blind<F>(&mut self, x: bigint::ElemDecoded<N>,
```

Comment:
BTW, the entire `bigint` API already assumes that we can't mix different (`N`, `P`, `Q`, ...) values, and that assumption is made specifically to avoid needing to compare moduli for equality on every multi-operand function call. That doesn't mean that blinding must do things the same way, of course.